connect/src/state.rs

```diff
@@ -194,6 +190,54 @@ impl ConnectState {
         }
     }
 
+    pub(crate) fn set_status(&mut self, status: &SpircPlayStatus) {
+        self.player.is_paused = matches!(
+            status,
+            SpircPlayStatus::LoadingPause { .. }
+                | SpircPlayStatus::Paused { .. }
+                | SpircPlayStatus::Stopped
+        );
+        self.player.is_buffering = matches!(
+            status,
+            SpircPlayStatus::LoadingPause { .. } | SpircPlayStatus::LoadingPlay { .. }
+        );
+        self.player.is_playing = matches!(
+            status,
+            SpircPlayStatus::LoadingPlay { .. } | SpircPlayStatus::Playing { .. }
+        );
+
+        debug!(
+            "updated connect play status playing: {}, paused: {}, buffering: {}",
+            self.player.is_playing, self.player.is_paused, self.player.is_buffering
+        );
+
+        if let Some(restrictions) = self.player.restrictions.as_mut() {
+            if self.player.is_playing && !self.player.is_paused {
+                restrictions.disallow_pausing_reasons.clear();
+                restrictions.disallow_resuming_reasons = vec!["not_paused".to_string()]
+            }
+
+            if self.player.is_paused && !self.player.is_playing {
+                restrictions.disallow_resuming_reasons.clear();
+                restrictions.disallow_pausing_reasons = vec!["not_playing".to_string()]
+            }
+        }
+    }
+
+    // todo: is there maybe a better or more efficient way to calculate the hash?
```

Comment:
Not 100% sure but my thinking is this: `VecDeque` implements `Hash` so you could make a smart pointer like `struct Queue(VecDeque)` and derive `Hash` on it. Then implement `Deref` on it so you can work with the `VecDeque` directly.